### PR TITLE
Callcenter

### PR DESCRIPTION
**Pull request by krl**

### CODE REVIEW

File: modules/callcenter/src/lib.rs [hunk 1]
```diff
@@ -26,11 +23,12 @@ static mut STATE: State<Callcenter> = State::new(Callcenter);
 
 impl Callcenter {
     pub fn query_counter(&self, counter_id: ModuleId) -> i64 {
-        dallo::query(counter_id, "read_value", ())
+        let result = uplink::query(counter_id, "read_value", ());
+        uplink::debug!("Quiring counter, got {}", result);
```
Comment (ureeves) [inline modules/callcenter/src/lib.rs]:
```suggestion
        uplink::debug!("Querying counter, got {}", result);
```

File: uplink/Cargo.toml [hunk 2]
```diff
@@ -8,6 +8,7 @@ license = "MPL-2.0"
 [dependencies]
 rkyv = { version = "0.7", default-features = false, features = ["size_32", "alloc"] }
 bytecheck = { version = "0.6", default-features = false }
+wee_alloc = "0.4.5"
```
Comment (ureeves) [inline uplink/Cargo.toml:11]: Both are fine, but we should decide if we want the usage of `wee_alloc` to be enforced in `uplink`, or if we want to leave it to contract writers. Keep in mind that there's an incentive to keep the bytecode small anyway.

File: uplink/src/helpers.rs [hunk 3]
```diff
@@ -25,12 +25,24 @@ where
     R: for<'a> Serialize<StandardBufSerializer<'a>>,
     F: Fn(A) -> R,
 {
+    crate::debug!("wrap query");
     with_arg_buf(|buf| {
         let slice = &buf[..arg_len as usize];
+
+        crate::debug!("slice {:?}", slice);
```
Comment (ureeves) [inline uplink/src/helpers.rs]: Should we keep these debug statements?

File: uplink/src/state.rs [hunk 4]
```diff
@@ -136,26 +125,28 @@ where
...
 
-    let name = raw.name();
-    let arg_len = raw.arg_bytes().len() as u32;
-    let ret_len = extern_query(mod_id, name, arg_len);
+    // let name = raw.name();
+    // let arg_len = raw.arg_bytes().len() as u32;
+    // let ret_len = unsafe { ext::q() };
 
-    with_arg_buf(|buf| RawResult::new(&buf[..ret_len as usize]))
+    // with_arg_buf(|buf| RawResult::new(&buf[..ret_len as usize]))
```
Comment (ureeves) [inline uplink/src/state.rs:140]: ?

File: uplink/src/state.rs [hunk 5]
```diff
@@ -166,13 +157,14 @@ where
...
-    with_arg_buf(|buf| {
-        let slice = &buf[..ret_len as usize];
-        let ret = unsafe { archived_root::<Ret>(slice) };
-        ret.deserialize(&mut Infallible).expect("Infallible")
-    })
+    // with_arg_buf(|buf| {
+    //     let slice = &buf[..ret_len as usize];
+    //     let ret = unsafe { archived_root::<Ret>(slice) };
+    //     ret.deserialize(&mut Infallible).expect("Infallible")
+    // })
```
Comment (ureeves) [inline uplink/src/state.rs:166]: ?x2

File: uplink/src/types.rs [hunk 6]
```diff
@@ -28,14 +27,15 @@ pub const MODULE_ID_BYTES: usize = 32;
...
+    CheckBytes,
     Deserialize,
     PartialOrd,
     Ord,
     Hash,
     Clone,
     Copy,
 )]
-#[archive_attr(derive(CheckBytes))]
+#[archive(as = "Self")]
```
Comment (ureeves) [inline uplink/src/types.rs:38]: This is indeed far better than an `ArchivedModuleId`

**Review by miloszm — CHANGES_REQUESTED** [review]
All looks good but make test fails, probably because produced wasm file is too big

**Review by miloszm — APPROVED** [review]

File: modules/micro/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+use uplink;
+
+static mut A: u32 = 42;
+
+#[no_mangle]
+unsafe fn change(to: u32) -> u32 {
+    let r = A;
+    A = to;
+    r
+}
```
Comment (ureeves) [inline modules/micro/src/lib.rs:20]: Does this work? Isn't it just a coincidence with the call convention?

**Review by ureeves — APPROVED** [review]
Small things here and there, but looks good.

**Pull request merged**